agent/src/runtime_connectors/podman_cli.rs

```diff
@@ -212,39 +281,41 @@ impl PodmanCli {
         Ok(id)
     }
 
-    pub async fn list_states_by_id(workload_id: &str) -> Result<Vec<ExecutionState>, String> {
-        let output = CliCommand::new(PODMAN_CMD)
-            .args(&[
-                "ps",
-                "--all",
-                "--filter",
-                &format!("id={workload_id}"),
-                "--format=json",
-            ])
-            .exec()
-            .await?;
-
-        let res: Vec<PodmanContainerInfo> = serde_json::from_str(&output)
-            .map_err(|err| format!("Could not parse podman output:{}", err))?;
-
-        Ok(res.into_iter().map(|x| x.into()).collect())
+    pub async fn list_states_by_id(workload_id: &str) -> Result<Option<ExecutionState>, String> {
+        let ps_result = LAST_PS_RESULT.get().await;
+        let all_containers_states = ps_result
+            .as_ref()
+            .container_states
+            .as_ref()
+            .map_err(|err| err.to_owned())?;
+        Ok(all_containers_states
+            .get(workload_id)
+            .map(ToOwned::to_owned))
     }
 
     pub async fn list_states_from_pods(pods: &[String]) -> Result<Vec<ContainerState>, String> {
-        if pods.is_empty() {
-            return Ok(vec![]);
-        }
-
-        let mut args = vec!["ps", "--all", "--format=json"];
-        let filters: Vec<String> = pods.iter().map(|p| format!("--filter=pod={}", p)).collect();
-        args.extend(filters.iter().map(|x| x as &str));
-
-        let output = CliCommand::new(PODMAN_CMD).args(&args).exec().await?;
+        let ps_result = LAST_PS_RESULT.get().await;
+        let all_pod_states = ps_result
+            .as_ref()
+            .pod_states
+            .as_ref()
+            .map_err(|err| err.to_owned())?;
+        Ok(pods
+            .iter()
+            .filter_map(|key| all_pod_states.get(key))
+            .flatten()
+            .map(|x| x.to_owned())
+            .collect())
```

Comment:
We have a bug here:
If a request is done to ask for the state of 2 pods for a workload and one of these pods is gone, the filter_map function would skip this input entry and return just the container states of the other pod. This would at the end result in a correct running state for the workload.
To reproduce you can create a workload with 2 pods and kill one using podman. The workload is still shown as running although one of its pods is not there anymore.

Reply:
The case of the whole pod being deleted is easy to fix:

```suggestion
        Ok(pods
            .iter()
            .flat_map(|key| {
                all_pod_states
                    .get(key)
                    .cloned()
                    .unwrap_or_else(|| vec![ContainerState::Unknown])
            })
            .collect())
```

But this will not work if only one container of the pod is removed. In this case the result would still be a "correct running state". I see not how we could fix this, as Podman only returns a list of pods when we create the workload, but not a list of containers.

Reply:
fixed